File: Leetcode/1626 Best Team With No Conflicts.py
from dataclasses import dataclass

class Solution:
    def bestTeamScore(self, scores, ages):
        score_ages = [ScoreAge(age, score) for age, score in zip(ages, scores)]
        score_ages.sort()

        best_scores = [score_age.score for score_age in score_ages]
        max_best_scores = float("-inf")

        for i in range(len(best_scores)):
            for j in range(i - 1, -1, -1):
                if score_ages[j].score <= score_ages[i].score:
                    best_scores[i] = max(best_scores[i], score_ages[i].score + best_scores[j])

            max_best_scores = max(max_best_scores, best_scores[i])

        return max_best_scores


@dataclass(order=True)
class ScoreAge:
    age: int
    score: int

File: Leetcode/1626 Best Team With No Conflicts.java
class Solution {
    public int bestTeamScore(int[] scores, int[] ages) {
        ScoreAge[] scoreAges = new ScoreAge[scores.length];
        for (int i = 0; i < scoreAges.length; ++i) {
            scoreAges[i] = new ScoreAge(scores[i], ages[i]);
        }
        Arrays.sort(scoreAges, ScoreAge.ageThenScore);
        
        int[] bestScores = new int[scores.length];
        int maxBestScores = Integer.MIN_VALUE;
        for (int i = 0; i < bestScores.length; ++i) {
            bestScores[i] = scoreAges[i].score;
            for (int j = i - 1; j >= 0; --j) {
                if (scoreAges[i].score >= scoreAges[j].score) {
                    bestScores[i] = Math.max(bestScores[i], scoreAges[i].score + bestScores[j]);
                }
            }
            maxBestScores = Math.max(maxBestScores, bestScores[i]);
        }
        return maxBestScores;
    }

    record ScoreAge(int score, int age) {
        private static final Comparator<ScoreAge> ageThenScore =
                Comparator.comparingInt(ScoreAge::age)
                          .thenComparingInt(ScoreAge::score);
    }
}